src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -231,26 +273,37 @@ public void OnRenderCompleted(long incomingBatchId, string errorMessageOrNull)
             // synchronizes calls to hub methods. That is, it won't issue more than one call to this method from the same hub
             // at the same time on different threads.
 
-            if (!UnacknowledgedRenderBatches.TryPeek(out var nextUnacknowledgedBatch) || incomingBatchId < nextUnacknowledgedBatch.BatchId)
+            if (!_unacknowledgedRenderBatches.TryPeek(out var nextUnacknowledgedBatch) || incomingBatchId < nextUnacknowledgedBatch.BatchId)
             {
                 Log.ReceivedDuplicateBatchAck(_logger, incomingBatchId);
             }
             else
             {
                 var lastBatchId = nextUnacknowledgedBatch.BatchId;
                 // Order is important here so that we don't prematurely dequeue the last nextUnacknowledgedBatch
-                while (UnacknowledgedRenderBatches.TryPeek(out nextUnacknowledgedBatch) && nextUnacknowledgedBatch.BatchId <= incomingBatchId)
+                while (_unacknowledgedRenderBatches.TryPeek(out nextUnacknowledgedBatch) && nextUnacknowledgedBatch.BatchId <= incomingBatchId)
                 {
                     lastBatchId = nextUnacknowledgedBatch.BatchId;
-                    UnacknowledgedRenderBatches.TryDequeue(out _);
+                    // At this point the queue is no longer full, we have at least emptied one slot, so we allow a further
```

Comment:
```suggestion
                    // At this point the queue is definitely not full, we have at least emptied one slot, so we allow a further
```

... since it wasn't necessarily full before